bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{subscriber::visitors::LocationVisitor, util::Flamegrapher, Error};
+
+use lazy_static::lazy_static;
+use parking_lot::RwLock;
+use tracing::{callsite, span, subscriber, Metadata, Subscriber};
+use tracing_subscriber::{layer::Context, registry::LookupSpan, Layer};
+
+use std::{
+    cell::Cell,
+    collections::{HashMap, HashSet},
+    fmt::{Display, Write as _},
+    fs::File,
+    io::{BufWriter, Write},
+    path::Path,
+    time::{Duration, Instant},
+};
+
+lazy_static! {
+    /// The initialisation time of this layer.
+    static ref START: Instant = Instant::now();
+}
+
+thread_local! {
+    /// The last event on this thread's current span stack.
+    static LAST_EVENT: Cell<Instant> = Cell::new(*START);
+}
+
+/// A [`tracing_subscriber::Layer`] for generating a flamegraph of tracked [`Span`](tracing::Span)s.
+///
+/// This layer registers an interest in the following [`Span`](tracing::Span) targets:
+///  - `bee::observe`: Generated by implementers of the [`Observe`](crate::Observe) trait.
+///  - `tokio::task`: Used internally by [`tokio`] to instrument [`task`](tokio::task)s, when compiling with
+/// `--cfg tokio_unstable` flags.
+///  - [`Span`](tracing::Span)s with the name `runtime.spawn`: used internally by newer versions of [`tokio`]
+/// to instrument [`task`](tokio::task)s, when compiling with `--cfg tokio_unstable` flags.
+pub struct FlamegraphLayer {
+    callsites: RwLock<HashSet<callsite::Identifier>>,
+    span_locations: RwLock<HashMap<span::Id, Option<SpanLocation>>>,
+    out_file: RwLock<BufWriter<File>>,
+}
+
+impl FlamegraphLayer {
+    pub(crate) fn new<P: AsRef<Path>>(stack_filename: P) -> Result<(Self, Flamegrapher), Error> {
+        let _ = *START;
```

Comment:
Why do you need to deref this?

Reply:
If it's not dereferenced here, it gets assigned the first time it is used elsewhere instead, which is innacurate (and can cause issues with `duration_since`.

Reply:
Oh I get it. I'd comment it just in case.